plugin-rust-agent/src/main/kotlin/jetbrains/buildServer/rust/AbstractToolProvider.kt

```diff
@@ -78,24 +90,35 @@ abstract class AbstractToolProvider(
             System.getenv(CargoConstants.ENV_CARGO_HOME)?.let {
                 this.add(it + File.separatorChar + "bin")
             }
-            this.addAll(StringUtil.splitHonorQuotes(System.getenv("PATH"), File.pathSeparatorChar))
-            add(System.getProperty("user.home") + File.separatorChar + ".cargo" + File.separatorChar + "bin")
+            System.getenv("PATH")?.let {
+                this.addAll(StringUtil.splitHonorQuotes(it, File.pathSeparatorChar))
+            }
+            System.getProperty("user.home")?.let {
+                this.add(it + File.separatorChar + ".cargo" + File.separatorChar + "bin")
+            }
         }
 
-        return paths.mapNotNull { File(it).listFiles() }
-                .flatMap { it.map { it.absolutePath } }
-                .filter { PATH_PATTERN.matches(it) }
-                .mapNotNull {
+        return paths.map { Path.of(it) }
+            .filter { Files.isDirectory(it) }
```

Comment:
Again, the filter is not needed, because by listing directories one has to handle errors anyway.